android/java/org/chromium/chrome/browser/playlist/kotlin/activity/PlaylistBaseActivity.kt

```diff
@@ -0,0 +1,167 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package org.chromium.chrome.browser.playlist.kotlin.activity
+
+import org.chromium.base.BraveFeatureList
+import org.chromium.base.BravePreferenceKeys
+import org.chromium.base.Log
+import org.chromium.base.supplier.OneshotSupplier
+import org.chromium.chrome.browser.app.BraveActivity
+import org.chromium.chrome.browser.flags.ChromeFeatureList
+import org.chromium.chrome.browser.init.ActivityProfileProvider
+import org.chromium.chrome.browser.init.AsyncInitializationActivity
+import org.chromium.chrome.browser.playlist.PlaylistServiceFactoryAndroid
+import org.chromium.chrome.browser.playlist.PlaylistServiceObserverImpl
+import org.chromium.chrome.browser.playlist.PlaylistServiceObserverImpl.PlaylistServiceObserverImplDelegate
+import org.chromium.chrome.browser.playlist.kotlin.enums.PlaylistOptionsEnum
+import org.chromium.chrome.browser.playlist.kotlin.listener.PlaylistItemOptionsListener
+import org.chromium.chrome.browser.playlist.kotlin.model.MoveOrCopyModel
+import org.chromium.chrome.browser.playlist.kotlin.model.PlaylistItemOptionModel
+import org.chromium.chrome.browser.playlist.kotlin.util.ConstantUtils
+import org.chromium.chrome.browser.playlist.kotlin.util.PlaylistUtils
+import org.chromium.chrome.browser.playlist.kotlin.view.bottomsheet.MoveOrCopyToPlaylistBottomSheet
+import org.chromium.chrome.browser.preferences.ChromeSharedPreferences
+import org.chromium.chrome.browser.profiles.ProfileProvider
+import org.chromium.chrome.browser.util.TabUtils
+import org.chromium.mojo.bindings.ConnectionErrorHandler
+import org.chromium.mojo.system.MojoException
+import org.chromium.playlist.mojom.PlaylistItem
+import org.chromium.playlist.mojom.PlaylistService
+
+abstract class PlaylistBaseActivity :
+    AsyncInitializationActivity(),
+    ConnectionErrorHandler,
+    PlaylistServiceObserverImplDelegate,
+    PlaylistItemOptionsListener {
+    companion object {
+        val TAG: String = "PlaylistBaseActivity"
+    }
+
+    protected var mPlaylistId = ConstantUtils.DEFAULT_PLAYLIST
+    protected var mPlaylistService: PlaylistService? = null
+    protected var mPlaylistServiceObserver: PlaylistServiceObserverImpl? = null
+
+    override fun onConnectionError(mojoException: MojoException) {
+        mPlaylistService?.close()
+        mPlaylistService = null
+        if (
+            ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_PLAYLIST) &&
+                ChromeSharedPreferences.getInstance()
+                    .readBoolean(BravePreferenceKeys.PREF_ENABLE_PLAYLIST, true)
+        ) {
+            initPlaylistService()
+        }
+    }
+
+    private fun initPlaylistService() {
+        if (mPlaylistService != null) {
+            mPlaylistService = null
+        }
+        mPlaylistService =
+            PlaylistServiceFactoryAndroid.getInstance()
+                .getPlaylistService(getProfileProviderSupplier().get()?.getOriginalProfile(), this)

```

Comment:
We should wait till profile is initialized like here https://github.com/chromium/chromium/blob/main/chrome/android/java/src/org/chromium/chrome/browser/download/DownloadManagerService.java#L1014

Reply:
ah in that case, we need to add profile observer and init service once it's fully initialized.